### DISCUSSION TITLE
Is `extract::State` really necessary?

### DISCUSSION THREAD

**Discussion by AlphaKeks** [tokio-rs/axum#2783] [Ideas]
State passed to [`Router::with_state()`](https://docs.rs/axum/0.7.5/axum/routing/struct.Router.html#method.with_state) must implement `Clone`, correct? Because of this, wouldn't it be possible to have a blanket implementation for `FromRequestParts<S> for S`, and get rid of the `State` extractor altogether?
```rust
#[async_trait]
impl<S> FromRequestParts<S> for S
where
    S: Clone,
{
    type Rejection = Infallible;

    async fn from_request_parts(_parts: &mut http::request::Parts, state: &S) -> Result<Self, Self::Rejection> {
        Ok(state.clone())
    }
}
```
I would think `State` does the same thing anyway, although I can't find an actual implementation for `FromRequestParts`, except for [this blanket impl](https://docs.rs/axum/0.7.5/axum/extract/struct.State.html#impl-FromRequestParts%3COuterState%3E-for-State%3CInnerState%3E).

In my own application I can implement `FromRequestParts<MyState> for MyState` successfully, but I don't know if another blanket impl would conflict with the existing one.

**Comment by jplatte** [comment]
This impl would conflict with all of the existing impls for extractors that implement `Clone`, so no, we can't do that.
